generate.py

```diff
@@ -0,0 +1,271 @@
+        n = n.parent
+        indents.pop()
+    # Indent
+    part = line.lstrip(' \t')
+    if len(part) != len(line):
+      indents.append(line[:len(line)-len(part)])
+    elif hang:
+      n = n.parent
+    hang = False
+    # Block syntax
+    if part.startswith(":"):
+      if n.data is None:
+        n.data = part[1:]
+      else:
+        n.data += "\n" + part[1:]
+      continue
+    # Node syntax
+    part, match = re_eat(r'^([A-Za-z0-9-.]{1,})', part)
+    if not match:
+      raise SyntaxError("expected node name")
+    n = n.append(Node(n, match[1]))
+    hang = True
+    # Direct value
+    if part.startswith(":"):
+      n.data = part[1:].strip(" \t")
+      continue
+    # Key-value pairs
+    part = part.lstrip(" \t")
+    while part != "":
+      part, match = re_eat(r'^([A-Za-z0-9-.]{1,})=', part)
+      if not match:
+        raise SyntaxError("expected pair key")
+      key = match[1]
+      part, match = re_eat(r'^([^\s"]+)|^"([^"]*)"', part)
+      if not match:
+        raise SyntaxError("expected pair value")
+      value = match[1]
+      n.append(Node(n, key, value))
+      part = part.lstrip(" \t")
+
+  return r
+
+# Base62 encoding
+b62charset = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+def b62encode(n):
+  s = ""
+  while n > 0:
+    s = b62charset[n % 62] + s
+    n //= 62
+  return s
+
+def manifest_url_hash(name):
+  return b62encode(int.from_bytes(sha256(name.encode('utf8')).digest(), "big"))[:4]
+
+def game_url_hash(hexdigest):
+  return b62encode(int.from_bytes(bytes.fromhex(hexdigest), "big"))[:8]
+
+def html_page(body: str):
+  return """<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>Preservation — byuu.org</title>
+</head>
+<body>
+  {}
+</body>
+</html>
+""".format(body)
+
+class Board:
+  def __init__(self, parent, manifest: Node):
+    self.name = manifest.text()
+    self.hash = manifest_url_hash(self.name)
+    self.manifest = manifest
+
+  def url(self):
+    return '{self.parent.url()}/{}'.format(self.hash)
+
+class BoardList:
+  def __init__(self, name, manifest: Node):
+    self.name = name
+    self.revision = manifest.path("database", "revision").text()
+    self.hash = manifest_url_hash(name)
+    self.boards = [Board(self, board) for board in manifest.elements('board')]
+    self.manifest = manifest
+
+  def html(self):
+    html = '<section><header>{}<span>{}</span></header>'.format(self.name, self.revision)
+    html += '<div>Total: {}</div>'.format(len(self.boards))
+    html += '<table><thead><tr><th>Name</th><tbody>'
+    for board in self.boards:
+      html += '<tr>'
+      html += '<td>{}</td>'.format(board.name)
+      html += '</tr>'
+    html += '</tbody></table></section>'
+    return html
+
+  def generate(self):
+    Path(os.path.join(OUTPATH, "boards", self.hash)).mkdir(parents=True, exist_ok=True)
+    Path(os.path.join(OUTPATH, "boards", self.hash, "index.html")).open("w").write(html_page(self.html()))
```

Comment:
You don't need to use `os.path.join()`, `Path` handles all that.

```suggestion
  def generate(self):
    outpath = Path(OUTPATH) / "boards" / self.hash
    outpath.mkdir(parents=True, exist_ok=True)
    (outpath / "index.html").open("w").write(html_page(self.html()))
```

Reply:
Thank you, i already realized this in my local copy but didn't push yet. (I refactored a lot of the HTML writing code.)

Reply:
\*blinks*

I'm sorry to maybe derail things and I get the need to not hardcode specific OS' path separators but what the frig is that syntax? Is that appending parts with *a `/` operator?*

Reply:
Probably, but imo the less ugly thing to do is `Path(OUTPATH, "boards", self.hash)` which also works just fine...

Reply:
Some people find overloading the division operator for "add directory component" to be a little too cute, and I probably wouldn't do that in my *own* code, but if it's in the standard library it feels like fair game.